Complete fourth stage

denshykk opened the pull request:
**Description**

You have created the foundation of our banking system. Now let's take the opportunity to deposit money into an account, make transfers and close an account if necessary.

Now your menu should look like this:

> 1. Balance
> 2. Add income
> 3. Do transfer
> 4. Close account
> 5. Log out
> 0. Exit

If the user asks for Balance, you should read the balance of the account from the database and output it into the console.

Add income item should allow us to deposit money to the account.

Do transfer item should allow transferring money to another account. You should handle the following errors:

If the user tries to transfer more money than he/she has, output: Not enough money!
If the user tries to transfer money to the same account, output the following message: You can't transfer money to the same account!
If the receiver's card number doesn’t pass the Luhn algorithm, you should output: Probably you made a mistake in the card number. Please try again!
If the receiver's card number doesn’t exist, you should output: Such a card does not exist.
If there is no error, ask the user how much money they want to transfer and make the transaction.
If the user chooses the Close an account item, you should delete that account from the database.

**Examples**
The symbol > represents the user input. Notice that it's not a part of the input.

Example 1:

```
1. Create an account
2. Log into account
0. Exit
>1

Your card has been created
Your card number:
4000009455296122
Your card PIN:
1961

1. Create an account
2. Log into account
0. Exit
>1

Your card has been created
Your card number:
4000003305160034
Your card PIN:
5639

1. Create an account
2. Log into account
0. Exit
>2

Enter your card number:
>4000009455296122
Enter your PIN:
>1961

You have successfully logged in!

1. Balance
2. Add income
3. Do transfer
4. Close account
5. Log out
0. Exit
>2

Enter income:
>10000
Income was added!

1. Balance
2. Add income
3. Do transfer
4. Close account
5. Log out
0. Exit
>1

Balance: 10000

1. Balance
2. Add income
3. Do transfer
4. Close account
5. Log out
0. Exit
>3

Transfer
Enter card number:
>4000003305160035
Probably you made a mistake in the card number. Please try again!

1. Balance
2. Add income
3. Do transfer
4. Close account
5. Log out
0. Exit
>3

Transfer
Enter card number:
>4000003305061034
Such a card does not exist.

1. Balance
2. Add income
3. Do transfer
4. Close account
5. Log out
0. Exit
>3

Transfer
Enter card number:
>4000003305160034
Enter how much money you want to transfer:
>15000
Not enough money!

1. Balance
2. Add income
3. Do transfer
4. Close account
5. Log out
0. Exit
>3

Transfer
Enter card number:
>4000003305160034
Enter how much money you want to transfer:
>5000
Success!

1. Balance
2. Add income
3. Do transfer
4. Close account
5. Log out
0. Exit
>1

Balance: 5000

1. Balance
2. Add income
3. Do transfer
4. Close account
5. Log out
0. Exit

>0
Bye!
```
```
Example 2:

1. Create an account
2. Log into account
0. Exit
>1

Your card has been created
Your card number:
4000007916053702
Your card PIN:
6263

1. Create an account
2. Log into account
0. Exit
>2

Enter your card number:
>4000007916053702
Enter your PIN:
>6263

You have successfully logged in!

1. Balance
2. Add income
3. Do transfer
4. Close account
5. Log out
0. Exit
>4

The account has been closed!

1. Create an account
2. Log into account
0. Exit
>2

Enter your card number:
>4000007916053702
Enter your PIN:
>6263

Wrong card number or PIN!

1. Create an account
2. Log into account
0. Exit
>0

Bye!
```

denshykk committed:
Complete fourth stage

denshykk committed:
Add default branch to switches; Manage warnings with Optionals

denshykk committed:
Add a new line in the eof of the pom.xml file

olegvovk96 commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,7 @@
+public class Main {
+
+    public static void main(String[] args) {
+        BankingSystem bankingSystem = new BankingSystem(args[1]);
```
What parameter I should use? Can you add it to the readme file?

olegvovk96 replied:
If there are no arguments it will be thrown - `ArrayIndexOutOfBoundsException`. Can you do validation?

olegvovk96 replied:
Why is the `BankingSystem` should aware of the DB name?

denshykk replied:
Updated readme file, made a validation for arguments and removed db name parameter in `BankingSystem` constructor.

olegvovk96 commented on src/main/java/dao/Dao.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.util.Optional;
+
+public interface Dao<T> {
```
Do you really need to use generic Dao?  You have only one Dao implementation. The method arguments names related to cards that make the interface more narrow in using.

(BTW I like that you know about it)

denshykk replied:
removed this interface

olegvovk96 commented on src/main/java/dao/AccountDao.java:
```diff
@@ -0,0 +1,132 @@
+import java.sql.Statement;
+import java.util.Optional;
+
+public class AccountDao implements Dao<Account> {
```
The `AccountDao` has too many obligations (create DB, table, connection, and queries to card table).
Can you refactor it?

denshykk replied:
Created a new class `DBCreator`, which is responsible for the db and the account `table` creation.

olegvovk96 commented on src/main/java/dao/AccountDao.java:
```diff
@@ -0,0 +1,132 @@
+public class AccountDao implements Dao<Account> {
+
+    private static final String CREATE_TABLE_QUERY = """
+            CREATE TABLE IF NOT EXISTS card(
```
In the domain, you are using the `Account` model, but the `card` table contains the same data. Can you normalize it?

denshykk replied:
Should I rename the `card` table to the `account`?

denshykk replied:
Renamed the `card` table to the `account`

olegvovk96 commented on src/main/java/dao/Account.java:
```diff
@@ -0,0 +1,6 @@
+
+import card.Card;
+
+public record Account(Card card, int balance) {
```
Can you move it to a `model` package?

denshykk replied:
Done.

olegvovk96 commented on src/main/java/card/CardGenerator.java:
```diff
@@ -0,0 +1,64 @@
+        return new Card(generateCard(), generatePIN());
+    }
+
+    public boolean isCorrectCardNumber(String cardNumber) {
```
The validation should be outside of card generation.

denshykk replied:
Exported validation to the `BankingSystem` class

olegvovk96 commented on src/main/java/card/CardGenerator.java:
```diff
@@ -0,0 +1,64 @@
+        final int bin = 400_000;
+        final int accountIdentifier = random.nextInt(900_000_000) + 100_000_000;
+        final int checkSum = luhnAlgorithm(bin, accountIdentifier);
+        return String.valueOf(bin) + accountIdentifier + checkSum;
```
Can you use `String.format()`?

denshykk replied:
Done.

olegvovk96 commented on src/main/java/card/CardGenerator.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    private int luhnAlgorithm(int bin, int accountIdentifier) {
+        String cardNumbersWithoutChecksum = ("" + bin + accountIdentifier);
```
Can you use `String.format()`?

denshykk replied:
Done.

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,163 @@
+    public BankingSystem(String dbName) {
+        dao = new AccountDao(dbName);
+        cardGenerator = new CardGenerator();
+        scanner = new Scanner(System.in);
```
Can you close the `scanner` in the end?

denshykk replied:
Closed.

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,163 @@
+                Your card PIN:
+                %s""", card.number(), card.pin());
+
+        System.out.println();
+        System.out.println();
```
Can you simplify it?
e.g.
```
System.out.print("\n".repeat(3));
```

denshykk replied:
Done

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,163 @@
+
+    private void logIntoAccount() {
+        System.out.println("\nEnter your card number:");
+        String inputtedCardNumber = scanner.next();
```
Can you simplify the field name?

denshykk replied:
Done.

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,163 @@
+        System.out.println("\nEnter your card number:");
+        String inputtedCardNumber = scanner.next();
+        System.out.println("Enter your PIN:");
+        String inputtedCardPIN = scanner.next();
```
Can you simplify the field name?

denshykk replied:
Done

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,163 @@
+        String inputtedCardNumber = scanner.next();
+        System.out.println("Enter your PIN:");
+        String inputtedCardPIN = scanner.next();
+        boolean isValid = accountValidation(inputtedCardNumber, inputtedCardPIN);
+
+        if (isValid) {
+            System.out.println("\nYou have successfully logged in!\n");
+            dao.get(inputtedCardNumber, inputtedCardPIN).ifPresent(account -> currentAccount = account);
+            accountMenu();
+        } else {
+            System.out.println("\nWrong card number or PIN!\n");
+        }
```
Can you use another Optional method `ifPresentOrElse()`? it should match better here.

denshykk replied:
Fixed.

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,163 @@
+        System.out.println("Enter income:");
+        int income = Integer.parseInt(scanner.next());
+        dao.update(currentAccount.card().number(), income);
+        dao.get(currentAccount.card().number(), currentAccount.card().pin()).ifPresent(account -> currentAccount = account);
```
Can you move it to a separate method? Looks like the same block is used in line 149.

olegvovk96 replied:
Looks like you only need to update `balance` in the `currentAccount`.

denshykk replied:
Made `Account` a simple POJO with balance setter instead of record

olegvovk96 commented:
Can you update the `Used technologies: Java, SQLite, **Gradle**` line in the Readme file?

denshykk committed:
Make changes according to comments in PR

- Update README.md
- Remove unneeded generic dao interface
- Do validation for wrong arguments
- Refactor AccountDao, create a new class DBCreator, which creates DB and a table "account"
- Change the name of the table from "card" to "account"
- Move classes to suitable packages
- Export card validation from CardGenerator to BankingSystem class
- Use String.format instead of concatenating ("+") strings
- Close the Scanner in the end
- Simplify printing a new line
- Simplify field names
- Use ifPresentOrElse() instead of ifPresent()
- Update only balance instead of creating a new Account instance

denshykk commented:
> Can you update the `Used technologies: Java, SQLite, **Gradle**` line in the Readme file?

Done, thank you)

olegvovk96 commented on pom.xml:
```diff
@@ -0,0 +1,24 @@
+        <dependency>
+            <groupId>org.xerial</groupId>
+            <artifactId>sqlite-jdbc</artifactId>
+            <version>3.8.11.2</version>
```
You are using the library version of the 2015 year. Can you update it?

denshykk replied:
done

olegvovk96 commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,23 @@
+                BankingSystem bankingSystem = new BankingSystem(accountDao);
+                bankingSystem.showMenu();
+            } else {
+                throw new IllegalArgumentException();
```
Can you remove the try-catch block? Instead of that:

- Add one more condition for validating allowed args length.
- Move message from lines 18-20 to the constructor of the IllegalArgumentException.

denshykk replied:
done

olegvovk96 commented on src/main/java/dao/AccountDao.java:
```diff
@@ -0,0 +1,87 @@
+            statement.setString(2, cardPIN);
+            ResultSet resultSet = statement.executeQuery();
+
+            int balance = resultSet.getInt("balance");
```
The balance may not be present, can you initialize it after the `resultSet.next()`?

denshykk replied:
fixed

olegvovk96 commented on src/main/java/dao/DBCreator.java:
```diff
@@ -0,0 +1,52 @@
+import java.sql.SQLException;
+import java.sql.Statement;
+
+public class DBCreator {
```
Can you rename it to `DBConfiguration` or `DataSourceConfiguration`?

denshykk replied:
Done

olegvovk96 commented on src/main/java/dao/AccountDao.java:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+
+    public void close() {
```
Can you move it to the `DBCreator`?

denshykk replied:
Done

olegvovk96 commented on src/main/java/model/CardGenerator.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.Random;
+
+public class CardGenerator {
```
I am not sure that should be kept in the `model` package.

denshykk replied:
Moved it to default package

olegvovk96 commented on src/main/java/model/Account.java:
```diff
@@ -0,0 +1,47 @@
+        this.balance = balance;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == this) return true;
+        if (obj == null || obj.getClass() != this.getClass()) return false;
+        var that = (Account) obj;
+        return Objects.equals(this.card, that.card) &&
+               this.balance == that.balance;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(card, balance);
+    }
+
+    @Override
+    public String toString() {
+        return "Account[" +
+               "card=" + card + ", " +
+               "balance=" + balance + ']';
+    }
```
Do you need to override all of these default methods?

denshykk replied:
Nope, removed

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,209 @@
+
+import java.util.Scanner;
+
+public class BankingSystem {
```
Can you use the default access modifier?

denshykk replied:
Changed

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,209 @@
+
+public class BankingSystem {
+
+    private final AccountDao dao;
```
Can you use the full name `accountDao` for better readability?

denshykk replied:
Done

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,209 @@
+
+            switch (input) {
+                case "0" -> {
+                    scanner.close();
+                    System.exit(0);
```
It should be the same as in lines 93-96. Can you move it to the common method?

denshykk replied:
Fixed

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,209 @@
+                .ifPresent(account -> currentAccount.setBalance(currentAccount.getBalance() + transferredMoney));
+    }
+
+    public boolean cardNumberValidation(String cardNumber) {
```
Can you use the private access modifier?

denshykk replied:
Done

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,209 @@
+        }
+    }
+
+    private boolean accountValidation(String inputtedCardNumber, String inputtedCardPIN) {
+        return dao.get(inputtedCardNumber, inputtedCardPIN).isPresent();
+    }
```
You duplicate call to the `dao.get`. Can you remove this method?

denshykk replied:
Removed

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,209 @@
+        String inCardNum = scanner.next();
+        System.out.println("Enter your PIN:");
+        String inCardPIN = scanner.next();
+        boolean isValid = accountValidation(inCardNum, inCardPIN);
+
+        if (isValid) {
+            System.out.println("\nYou have successfully logged in!\n");
+            dao.get(inCardNum, inCardPIN).ifPresentOrElse(account -> currentAccount = account,
+                    () -> System.out.println("\nWrong card number or PIN!\n"));
+            accountMenu();
+        }
```
You don't need to call the `dao.get` two times. Remove `if statement` and fix `ifPresentOrElse` logic.

denshykk replied:
Remove if statement and fixed ifPresentOrElse logic

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,209 @@
+                           "Enter card number:");
+        String recipientCardNumber = scanner.next();
+
+        if (recipientCardNumber.equals(currentAccount.getCard().number())) {
+            System.out.println("You can't transfer money to the same account!\n");
+            return;
+        }
+
+        if (!cardNumberValidation(recipientCardNumber)) {
+            System.out.println("Probably you made a mistake in the card number. Please try again!\n");
+            return;
+        }
+
+        if (!dao.contains(recipientCardNumber)) {
+            System.out.println("Such a card does not exist.\n");
+            return;
+        }
```
Can you move it to a separate method?

denshykk replied:
Moved.

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,209 @@
+            }
+
+            dao.update(recipientCardNumber, moneyToTransfer);
+            updateAccount(currentAccount.getCard().number(), -moneyToTransfer);
```
Can you put `-moneyToTransfer` in a new variable and call it with the prefix `negate`?

denshykk replied:
Created a new variable negateRemittance and initialized it with negative moneyToTransfer value

denshykk committed:
- Update SQLite library version
- Replace try-catch in the Main class to if with arguments array length
  check
- Relocate getting balance to the if in order to escape exception if it
  won't be initialized
- Rename DBCreator to DBConfiguration to get better readibility
- Move close method to the DBConfiguration class
- Correct packages structure
- Removed unnecessary methods
- Change access modifier of BankingSystem class to default
- Rename reference variable dao to account dao
- Move the logic of program's end to the separate method named
  stopApplication()
- Change access modifier of cardNumberValidation() method to private
- Remove accountValidation() method
- Move recipient card number validation logic to the
  recipientCredentialsValidation() method
- Fix ifPresentOrElse logic
- Create a new variable negateRemittance and pass it in order to subtract money, that was transfered to another account
  specific account instead of passing -moneyToTransfer

olegvovk96 commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,19 @@
+public class Main {
+
+    public static void main(String[] args) {
+        if (args[0].equals("-fileName") && args.length == 2) {
```
If args is empty, the `ArrayIndexOutOfBoundsException` still will be thrown. You can put the length check in the first place.

denshykk replied:
Fixed

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,214 @@
+        System.out.println("Enter your PIN:");
+        String inCardPIN = scanner.next();
+
+        accountDao.get(inCardNum, inCardPIN).ifPresentOrElse(account -> {
+                    currentAccount = account;
+                    System.out.println("\nYou have successfully logged in!\n");
+                    accountMenu();
+                },
```
That's looking better now, good job! But it can look even better)
Can you please move it to a separate method
```
account -> {
       currentAccount = account;
       System.out.println("\nYou have successfully logged in!\n");
       accountMenu();
}
```
then call it using method reference. It should look something like this
```
accountDao.get(inCardNum, inCardPIN).ifPresentOrElse(
             this::doSomething,
             () -> System.out.println("\nWrong card number or PIN!\n"));
```

denshykk replied:
Done

olegvovk96 commented on src/main/java/BankingSystem.java:
```diff
@@ -0,0 +1,214 @@
+            doTransfer();
+        }
+    }
+
```
Can you remove the empty line?

denshykk replied:
Done

olegvovk96 commented on src/main/java/CardGenerator.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.Random;
+
+public class CardGenerator {
```
Can you use the default access modifier?

denshykk replied:
Fixed

olegvovk96 commented on src/main/java/dao/DBConfiguration.java:
```diff
@@ -0,0 +1,60 @@
+import java.sql.SQLException;
+import java.sql.Statement;
+
+public class DBConfiguration {
```
Can you move this class to a `configuration` package?

denshykk replied:
Done

denshykk committed:
- Changed the argument name check with arguments array length check places to avoid ArrayIndexOutOfBoundsException
- Extract logic if logging into the account was successful and pass
  methods reference in ifPresent() method
- Remove empty line
- Change access modifier of CardGenerator class to default
- Move DBConfiguration class to the configuration package

olegvovk96 commented on src/main/java/configuration/DBConfiguration.java:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+
+    public void close() {
```
Can you rename this method to `closeConnection`?

denshykk replied:
Done.

denshykk committed:
Rename close() method to closeConnection()

olegvovk96 commented:
Awesome work!
